README


This archive contains the distribution Games-Dice-Loaded,
version 0.001:

  Perl extension to simulate rolling loaded dice

This software is copyright (c) 2012 by Miles Gould.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Games-Dice-Loaded

0.001     2012-01-11 19:17:28 Europe/London

	- Some tests to see that we do actually produce the desired distribution

	- Added Test::Probability, a module for performing these tests

	- More use of statistical language in the docs

	- die() rather than croak()ing on internal consistency errors.

0.001_01  2012-01-10 23:34:43 Europe/London

	- Implementation of the algorithm.

	- Some simple sanity checks.


lib/Test/Probability.pm
package Test::Probability;
{
  $Test::Probability::VERSION = '0.001';
}

use strict;
use warnings;

require Exporter;
our @ISA = qw/Exporter/;
our @EXPORT = qw/dist_ok/;

require AutoLoader; # Statistics::Distributions should do this, but doesn't.
use Statistics::Distributions 1.02 qw/chisqrdistr/;
use Test::More;
use List::Util qw/sum/;
use List::MoreUtils qw/pairwise/;
use Carp;

sub chisq {
	my ($observed, $expected) = @_;
	return sum(pairwise {
		my $denom = $b > 0.5 ? $b : 0.5;
		($a - $b)**2 / $denom;
	} @$observed, @$expected);
}

sub fits_distribution {
	my ($observed, $dist, $confidence) = @_;
	croak "First arg not an array or arrayref" unless ref $observed eq 'ARRAY';
	croak "Second arg not an array or arrayref" unless ref $dist eq 'ARRAY';
	croak "Size of observed domain does not match size of distribution domain"
		unless scalar(@$observed) == scalar(@$dist);
	my $observations = sum @$observed;
	my $scalefactor = $observations / (sum @$dist);
	my @expected = map { $_ * $scalefactor } @$dist;
	my $chisq = chisq($observed, \@expected);
	my $dof = @$dist - 1;	# degrees of freedom
	my $threshold = chisqrdistr($dof, 1 - $confidence);
	return $chisq <= $threshold;
}

sub dist_ok {
	my ($observed, $dist, $confidence, $message) = @_;
	ok(fits_distribution($observed, $dist, $confidence), $message);
}

1;

__END__

=head1 NAME

Test::Probability - test if results are distributed correctly

=head1 SYNOPSIS

  use Test::Probability;

  my @results;
  for (0 .. 1000) {
	$results[rand_fn()]++;
  }
  my @probabilities = (0.5, 0.2, 0.3);
  my $confidence = 0.9;
  dist_ok(@results, @probabilities, $confidence,
    "results match expected with confidence $confidence");
  
=head1 DESCRIPTION

Does your random-number generating function output the distribution you expect?
Now you can find out! This module performs a chi-squared test at the specified
confidence interval.

=head1 EXPORTS

=item dist_ok

  dist_ok(@results, @probabilities, $confidence,
    "results match probabilities with confidence $confidence");
  
=head1 AUTHOR

Miles Gould, E<lt>mgould@cpan.orgE<gt>

=head1 CONTRIBUTING

This module is currently distributed along with Games::Dice::Loaded.
Please fork
L<the GitHub repository|http://github.com/pozorvlak/Games-Dice-Loaded>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Miles Gould

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=head1 SEE ALSO

=over

=item L<Statistics::ChiSquared> (only handles even distributions).

=item L<Statistics::Distributions> (used internally by this module).

=back

=cut


lib/Games/Dice/Loaded.pm
package Games::Dice::Loaded;
{
  $Games::Dice::Loaded::VERSION = '0.001';
}
use Moose 2.0300;
use List::Util qw/max sum/;

# ABSTRACT: Perl extension to simulate rolling loaded dice

# Keith Schwarz's article is lovely and has lots of pretty diagrams and proofs,
# but unfortunately it's also very long. Here's the tl;dr:

# Draw a bar chart of the probabilities of landing on the various faces, then
# throw darts at it (by picking X and Y coordinates uniformly at random). If
# you hit a bar with your dart, choose that face. This works OK, but has very
# bad worst-case behaviour; fortunately, it's possible to cut up the taller
# bars and stack them on top of the shorter bars in such a way that the area
# covered is exactly a (1/n) \* n rectangle. Constructing this rectangular
# "dartboard" can be done in O(n) time, by maintaining a list of short (less
# than average height) bars and a list of long bars; add the next short bar to
# the dartboard, then take enough of the next long bar to fill that slice up to
# the top. Add the index of the long bar to the relevant entry of the "alias
# table", then put the remainder of the long bar back into either the list of
# short bars or the list of long bars, depending on how long it now is.

# Once we've done this, simulating a dice roll can be done in O(1) time:
# Generate the dart's coordinates; which vertical slice did the dart land in,
# and is it in the shorter bar on the bottom or the "alias" that's been stacked
# above it?

# Heights of the lower halves of the strips
has 'dartboard' => ( is => 'ro', isa => 'ArrayRef' );
# Identities of the upper halves of the strips
has 'aliases' => ( is => 'ro', isa => 'ArrayRef' );
has 'num_faces' => ( is => 'ro', isa => 'Num' );

# Construct the dartboard and alias table
around BUILDARGS => sub {
	my $orig = shift;
	my $class = shift;
	# scale so average weight is 1
	my @weights = @_;
	my $n = scalar @weights;
	my $scalefactor = $n / sum(@weights);
	my $i = 0;
	@weights = map { [$i++, $scalefactor * $_] } @weights; 
	my @small = grep { $_->[1] < 1 } @weights;
	my @large = grep { $_->[1] >= 1 } @weights;
	my @dartboard; my @aliases;
	while ((@small > 0) && (@large > 0)) {
		my ($small_id, $small_p) = @{pop @small};
		my ($large_id, $large_p) = @{pop @large};
		$dartboard[$small_id] = $small_p;
		$aliases[$small_id] = $large_id;
		$large_p = $small_p + $large_p - 1;
		if ($large_p >= 1) {
			push @large, [$large_id, $large_p];
		} else {
			push @small, [$large_id, $large_p];
		}
	}
	for my $unused (@small, @large) {
		$dartboard[$unused->[0]] = 1;
		$aliases[$unused->[0]] = $unused->[0];
	}
	for my $face (0 .. $n - 1) {
		my $d = $dartboard[$face];
		die("Undefined dartboard for face $face") unless defined $d;
		die("Height $d too large for face $face") unless $d <= 1;
		die("Height $d too small for face $face") unless $d >= 0;
	}
	return $class->$orig(
		dartboard => \@dartboard,
		aliases => \@aliases,
		num_faces => $n,
	);
};

# Roll the die
sub roll {
	my ($self) = @_;
	my $face = int(rand $self->num_faces);
	my $height = rand 1;
	my @dartboard = @{$self->dartboard()};
	die("Dartboard undefined for face $face")
		unless defined $dartboard[$face];
	if ($height > $dartboard[$face]) {
		my @aliases = @{$self->aliases};
		return $aliases[$face] + 1;
	} else {
		return $face + 1;
	}
}

*sample = \&roll;

1;

__END__

=head1 NAME

Games::Dice::Loaded - Simulate rolling loaded dice

=head1 SYNOPSIS

  use Games::Dice::Loaded;

  my $die = Games::Dice::Loaded->new(1/6, 1/6, 1/2, 1/12, 1/12);
  my $result = $die->roll();

  my $fair_d4 = Games::Dice::Loaded->new(1, 1, 1, 1);
  $result = $fair_d4->roll();

=head1 DESCRIPTION

C<Games::Dice::Loaded> allows you to simulate rolling arbitrarily-weighted dice
with arbitrary numbers of faces - or, more formally, to sample any discrete
probability distribution which may take only finitely many values. It does this
using Vose's elegant I<alias method>, which is described in Keith Schwarz's
article L<Darts, Dice, and Coins: Sampling from a Discrete
Distribution|http://www.keithschwarz.com/darts-dice-coins/>.

=head1 METHODS

=over

=item new()

Constructor. Takes as arguments the probabilities of rolling each "face". If
the weights given do not sum to 1, they are scaled so that they do. This method
constructs the alias table, in O(num_faces) time.

=item roll()

Roll the die. Takes no arguments, returns a number in the range 1 .. num_faces.
Takes O(1) time.

=item sample()

Synonym for C<roll()>.

=item num_faces()

The number of faces on the die. More formally, the size of the discrete random
variable's domain. Read-only.

=back

=head1 AUTHOR

Miles Gould, E<lt>mgould@cpan.orgE<gt>

=head1 CONTRIBUTING

Please fork
L<the GitHub repository|http://github.com/pozorvlak/Games-Dice-Loaded>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Miles Gould

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.4 or,
at your option, any later version of Perl 5 you may have available.

=head1 SEE ALSO

Perl modules for rolling dice:
L<Games::Dice>,
L<Games::Dice::Advanced>,
L<Bot::BasicBot::Pluggable::Module::Dice>,
L<random>.

A Perl module for calculating probability distributions for dice rolls:
L<Games::Dice::Probability>.

Descriptions of the alias method:

=over

=item Michael Vose, L<A Linear Algorithm For Generating Random Numbers with a Given Distribution|http://web.eecs.utk.edu/~vose/Publications/random.pdf>

=item Keith Schwarz, L<Darts, Dice, and Coins: Sampling from a Discrete
Distribution|http://www.keithschwarz.com/darts-dice-coins/>

=item L<Data structure for loaded dice?|http://stackoverflow.com/questions/5027757/data-structure-for-loaded-dice> on StackOverflow

=item L<Wikipedia article|http://en.wikipedia.org/wiki/Alias_method>

=back

=cut


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl extension to simulate rolling loaded dice",
  "AUTHOR" => "Miles Gould <miles\@assyrian.org.uk>",
  "BUILD_REQUIRES" => {
    "Test::More" => "0.88"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Games-Dice-Loaded",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Games::Dice::Loaded",
  "PREREQ_PM" => {
    "AutoLoader" => 0,
    "Carp" => 0,
    "Exporter" => 0,
    "List::MoreUtils" => 0,
    "List::Util" => 0,
    "Moose" => "2.0300",
    "Statistics::Distributions" => "1.02",
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



